### PR TITLE
add request forwarding for telemetry

### PR DESCRIPTION
**Pull request by wconti27**
This PR adds two middlewares to the Test-Agent. The first middleware stores the request data and headers in Test-Agent memory. The second middleware forwards requests to an optional end destination (such as a real DD Agent) for any requests from the `/traces`, `/stats`, or `/telemetry` endpoints.

### CODE REVIEW

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]

File: ddapm_test_agent/agent.py [hunk 1]
```diff
@@ -112,6 +112,48 @@ async def session_token_middleware(request: Request, handler: _Handler) -> web.R
     return await handler(request)
 
 
+async def _forward_request(request_data, headers, agent_url):
+    async with ClientSession() as session:
+        async with session.post(
+            f"{agent_url}/v0.4/traces",
```
[the remark this answers is not in the document]
Comment (wconti27) [reply ddapm_test_agent/agent.py:118]: good catch

File: ddapm_test_agent/agent.py [hunk 2]
```diff
@@ -652,6 +642,64 @@ async def handle_trace_analyze(self, request: Request) -> web.Response:
...
+        if "X-Datadog-Agent-Proxy-Disabled" in headers:
+            request["_proxy_to_agent"] = headers.pop("X-Datadog-Agent-Proxy-Disabled").lower() != "true"
+
+        if "X-Datadog-Proxy-Port" in headers:
+            port = headers.pop("X-Datadog-Proxy-Port")
+            request.app["agent_url"] = update_trace_agent_port(request.app["agent_url"], new_port=port)
+            log.info("Found port in headers, new trace agent URL is: {}".format(request.app["agent_url"]))
+
+        request["_headers"] = headers
+        # Call the original handlerx
```
Comment (Kyle-Verhoog) [inline ddapm_test_agent/agent.py:669]:
```suggestion
        # Call the original handler
```

File: ddapm_test_agent/agent.py [hunk 3]
```diff
@@ -652,6 +642,64 @@ async def handle_trace_analyze(self, request: Request) -> web.Response:
...
+
+        if "X-Datadog-Trace-Env-Variables" in headers:
+            var_string = headers.pop("X-Datadog-Trace-Env-Variables")
+            env_vars = {
+                key.strip(): value.strip() for key, value in (pair.split("=") for pair in var_string.split(","))
+            }
+            request["_dd_trace_env_variables"] = env_vars
+
+        if "X-Datadog-Agent-Proxy-Disabled" in headers:
+            request["_proxy_to_agent"] = headers.pop("X-Datadog-Agent-Proxy-Disabled").lower() != "true"
```
Comment (Kyle-Verhoog) [inline ddapm_test_agent/agent.py:661]: this check should be in the proxy codepath, not the request storing

File: ddapm_test_agent/agent.py [hunk 4]
```diff
@@ -652,6 +642,64 @@ async def handle_trace_analyze(self, request: Request) -> web.Response:
...
+        else:
+            # Call the original handler and do nothing
+            return await handler(request)
+
+    async def _forward_request_to_agent(self, request: Request, handler: _Handler) -> web.Response:
+        """Forward all requests to the agent_url if set."""
+        data = self._request_data(request)
+        headers = request["_headers"]
+        agent_url = request.app["agent_url"]
+        proxy_to_agent = request.get("_proxy_to_agent", True)
```
Comment (Kyle-Verhoog) [inline ddapm_test_agent/agent.py:686]: why are we changing the condition for proxying to the agent? Before it was if there was an agent url detected (and then checking the header). 

```
        proxy_to_agent = agent_url != ""

        if "X-Datadog-Agent-Proxy-Disabled" in headers:
            proxy_to_agent = headers.pop("X-Datadog-Agent-Proxy-Disabled").lower() != "true"
```

can we maintain this functionality?

**Review by Kyle-Verhoog — COMMENTED** [review]
This change needs a release note. Also it would be great if we could have tests for the request storing logic to ensure it works.

Otherwise, I really like the middleware implementation 👍

File: ddapm_test_agent/agent.py (hunk 3 above)
Comment (wconti27) [reply ddapm_test_agent/agent.py:661]: IMO I don't agree. I think centrally locating any header transformations / checks is useful. To get the same functionality as before, I would have to do this header check within each of the forward endpoints (traces, stats, telemetry).  I added the check to make sure agent_url isn't empty. Should be the same functionality as before.

File: ddapm_test_agent/agent.py (hunk 4 above)
Comment (wconti27) [reply ddapm_test_agent/agent.py:686]: This functionality is maintained

File: ddapm_test_agent/agent.py (hunk 3 above)
Comment (Kyle-Verhoog) [reply ddapm_test_agent/agent.py]: wouldn't it be centralized in the proxy codepath/middleware?

File: ddapm_test_agent/agent.py [hunk 5]
```diff
@@ -134,6 +177,14 @@ def __init__(self):
         # Token to be used if running test cases synchronously
         self._requests: List[Request] = []
         self._rc_server = RemoteConfigServer()
+        self._trace_failures: List[str] = []
```
Comment (Kyle-Verhoog) [inline ddapm_test_agent/agent.py:180]:
```suggestion
```

unrelated change?

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]
